### PR TITLE
[🧪] NT-888 Campaign details experiment on Campaign details page

### PR DESCRIPTION
**Pull request by eoji**
# 📲 What
Displaying native_project_page_campaign_details variants on live and not backed Campaign details screen.

# 🤔 Why
🧬

# 🛠 How
- Added `ActivityRequestCodes.SHOW_REWARDS` when starting `CampaignDetailsActivity`
- Added `IntentKey.PROJECT_DATA` for passing `ProjectData` to started Activity
- `CampaignDetails`
  - `CampaignDetailsActivity` is now started with `ProjectData`
  - Added faux pledge container `campaign_details_pledge_container`
  - `CampaignDetailsViewModel`
    - Added `pledgeActionButtonClicked` input that is called when the user clicks the pledge action button.
    - Added `goBackToProject` output that emits when we should dismiss this screen and show the rewards.
    - Added `pledgeContainerIsVisible` output that emits a `Boolean` determining if the faux pledge container should be visible.
    - Tests
- `Project`
  - `ProjectActivity.startCampaignWebViewActivity` now takes in a `[ProjectData]` and starts `CampaignDetailsActivity` for a result of `SHOW_REWARDS`
  - `ProjectViewModel`
    - `startCampaignWebViewActivity` output now emits `ProjectData`
    - `expandPledgeSheet` emits when an `OK` `SHOW_REWARDS` result is received.
    - Tests

# 👀 See
| Variant 2 🐛 | Variant 1 🐛 | Control 🐛 |
| --- | --- | --- |
| screenshot-2020-02-20_152913 | screenshot-2020-02-20_152903 | screenshot-2020-02-20_152903 |
| device-2020-02-20-134307 2020-02-20 13_46_55 |

# 📋 QA
Get yourself into variant-2 of `native_project_page_campaign_details` and check out the campaign details!
The faux pledge container should only be visible on live and unbacked projects.

# Story 📖
[NT-888]

[NT-888]: https://kickstarter.atlassian.net/browse/NT-888

### CODE REVIEW

File: app/src/main/java/com/kickstarter/libs/ActivityRequestCodes.java [hunk 1]
```diff
@@ -8,4 +8,5 @@ private ActivityRequestCodes() {}
   public final static int CHECKOUT_ACTIVITY_WALLET_CHANGE_REQUEST = 592_11;
   public final static int CHECKOUT_ACTIVITY_WALLET_OBTAINED_FULL = 592_12;
   public final static int SAVE_NEW_PAYMENT_METHOD = 592_13;
+  public final static int SHOW_REWARDS = 592_14;
```
Comment (justinswart) [inline app/src/main/java/com/kickstarter/libs/ActivityRequestCodes.java]: What's a request code?

File: app/src/main/java/com/kickstarter/viewmodels/CampaignDetailsViewModel.kt [hunk 2]
```diff
@@ -1,42 +1,79 @@
...
+        /** Call when the user clicks the pledge action button. */
+        fun pledgeActionButtonClicked()
+    }
 
     interface Outputs {
+        /** Emits when we should return to the [com.kickstarter.ui.activities.ProjectActivity] with the rewards visible. */
+        fun goBackToProject(): Observable<Void>
+
+        /** Emits a boolean determining if the faux pledge container is visible. */
+        fun pledgeContainerIsVisible(): Observable<Boolean>
```
Comment (justinswart) [inline app/src/main/java/com/kickstarter/viewmodels/CampaignDetailsViewModel.kt]: Ooh, not isGone? 😄

File: app/src/test/java/com/kickstarter/viewmodels/CampaignDetailsViewModelTest.kt [hunk 3]
```diff
@@ -3,24 +3,98 @@ package com.kickstarter.viewmodels
...
+    @Test
+    fun testGoBackToProject() {
+        setUpEnvironment(environment(), ProjectDataFactory.project(ProjectFactory.project()))
+
+        this.vm.inputs.pledgeActionButtonClicked()
+        this.goBackToProject.assertValueCount(1)
+    }
+
+    @Test
+    fun testPledgeContainerIsVisible_whenProjectIsLiveNotBacked_control() {
```
Comment (justinswart) [inline app/src/test/java/com/kickstarter/viewmodels/CampaignDetailsViewModelTest.kt]: Couple questions about these tests:

1) This test appears to be testing the container visibility but asserts after `this.vm.inputs.pledgeActionButtonClicked()`, does the button click affect its visibility?
2) What do you think about `testPledgeContainerIsNotVisible` or `testPledgeContainerVisibility`? Usually I read the test name almost as an assertion that it's testing that _this thing is this_, so in this case I initially thought it was testing that the container _is_ visible whereas it's actually the opposite?

File: app/src/test/java/com/kickstarter/viewmodels/CampaignDetailsViewModelTest.kt [hunk 4]
```diff
@@ -3,24 +3,98 @@ package com.kickstarter.viewmodels
...
+                })
+                .build()
+        setUpEnvironment(environment, ProjectDataFactory.project(ProjectFactory.project()))
+
+        this.vm.inputs.pledgeActionButtonClicked()
+        this.pledgeContainerIsVisible.assertValue(true)
+    }
+
+    @Test
+    fun testPledgeContainerIsVisible_whenProjectIsNotLive() {
```
Comment (justinswart) [inline app/src/test/java/com/kickstarter/viewmodels/CampaignDetailsViewModelTest.kt]: For these last two I'm guessing the container is not visible by default but not sure if we want to explicitly emit `false` in these cases 🤷‍♂

File: app/src/main/java/com/kickstarter/libs/ActivityRequestCodes.java (hunk 1 above)
Comment (eoji) [reply app/src/main/java/com/kickstarter/libs/ActivityRequestCodes.java]: If you want to get a result back from a screen you started, you need to pass in a request code `Int`. https://developer.android.com/training/basics/intents/result#StartActivity. So if I want the `ProjectActivity` to know that the `CampaignDetailsActivity`'s pledge action button was clicked, I need to use a unique request code.

File: app/src/test/java/com/kickstarter/viewmodels/CampaignDetailsViewModelTest.kt (hunk 3 above)
Comment (eoji) [reply app/src/test/java/com/kickstarter/viewmodels/CampaignDetailsViewModelTest.kt]: 1. No, it's copy pasta I need to update 😅 
2. We've been trying to name tests `testOutputName_whenCondition`

File: app/src/test/java/com/kickstarter/viewmodels/CampaignDetailsViewModelTest.kt (hunk 4 above)
Comment (eoji) [reply app/src/test/java/com/kickstarter/viewmodels/CampaignDetailsViewModelTest.kt]: It is initially hidden but I can make it emit in those cases 👍

File: app/src/main/java/com/kickstarter/viewmodels/CampaignDetailsViewModel.kt (hunk 2 above)
Comment (eoji) [reply app/src/main/java/com/kickstarter/viewmodels/CampaignDetailsViewModel.kt]: I think `isVisible` is less confusing but should I do `isGone` to be consistently confusing?

Comment (justinswart) [reply app/src/main/java/com/kickstarter/viewmodels/CampaignDetailsViewModel.kt]: Yeah, in our case we just match it to the property it's setting which is `isHidden`. Whatever works!

**Pull request merged**